File: 202309/1658.go
func minOperations(nums []int, x int) int {
	n := len(nums)
	tot := 0
	for _, a := range nums {
		tot += a
	}
	tot -= x
	j := 0
	curr := 0
	ans := n + 1
	for i := 0; i < n; i++ {
		curr += nums[i]
		for j <= i && curr > tot {
			curr -= nums[j]
			j++
		}
		if curr == tot {
			ans = min(ans, n-(i-j+1))
		}
	}
	if ans == n+1 {
		return -1
	}
	return ans
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: 202609/1658.cpp
class Solution {
public:
  int minOperations(vector<int> &nums, int x) {
    int curr = 0;
    unordered_map<int, int> m;
    m[0] = 0;
    for (int i = 0; i < nums.size(); ++i) {
      curr += nums[i];
      if (curr <= x)
        m[curr] = i + 1;
      else
        break;
    }
    int ans = nums.size() + 1;
    if (m.find(x) != m.end()) {
      ans = m[x];
    }
    curr = 0;
    for (int i = nums.size() - 1; i >= 0; --i) {
      curr += nums[i];
      if (m.find(x - curr) != m.end()) {
        ans = min(ans, (int)nums.size() - i + m[x - curr]);
      }
    }
    if (ans > nums.size())
      return -1;
    return ans;
  }
};
